File: src/CVNBot/ProjectList.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Xml;
using log4net;

namespace CVNBot
{
    class ProjectList : SortedList
    {
        ILog logger = LogManager.GetLogger("CVNBot.ProjectList");

        public string fnProjectsXML;
        public string currentBatchReloadChannel = "";

        /// <summary>
        /// Dumps all Projects to an XML file (Projects.xml)
        /// </summary>
        void DumpToFile()
        {
            logger.Info("Saving configuration to " + fnProjectsXML);
            StreamWriter sw = new StreamWriter(fnProjectsXML);
            sw.WriteLine("<projects>");
            foreach (DictionaryEntry dicent in this)
            {
                Project prj = (Project)dicent.Value;
                // Get each Project's details and append it to the XML file
                sw.WriteLine(prj.DumpProjectDetails());
            }
            sw.WriteLine("</projects>");
            sw.Flush();
            sw.Close();
        }

        /// <summary>
        /// Loads and initializes Projects from an XML file (Projects.xml)
        /// </summary>
        public void LoadFromFile()
        {
            logger.Info("Reading projects from " + fnProjectsXML);
            XmlDocument doc = new XmlDocument();
            doc.Load(fnProjectsXML);
            XmlNode parentnode = doc.FirstChild;
            for (int i = 0; i < parentnode.ChildNodes.Count; i++)
            {
                string prjDefinition = "<project>" + parentnode.ChildNodes[i].InnerXml + "</project>";
                Project prj = new Project();
                prj.ReadProjectDetails(prjDefinition);
                this.Add(prj.projectName, prj);
            }
        }

        /// <summary>
        /// Adds a new Project to the ProjectList. Remember to dump the configuration afterwards by calling dumpToFile()
        /// </summary>
        /// <param name="projectName">Name of the project (e.g., en.wikipedia) to add</param>
        /// <param name="interwiki">Interwiki link (e.g., it:s: -- can be empty string)</param>
        /// <param name="lang">ISO 639 lang code (e.g. "it", "es" Optional) </param>
        public void AddNewProject(string projectName, string interwiki, string lang = "")
        {
            if (interwiki == "")
            {
                // Try to guess interwiki

                if (!projectName.Contains("."))
                {
                    // Cannot guess; probably something like "mediawiki"
                    throw new Exception((String)Program.msgs["20004"]);
                }

                string langPortion = projectName.Split(new char[] { '.' }, 2)[0];
                string projPortion = projectName.Split(new char[] { '.' }, 2)[1];
                switch (projPortion)
                {
                    case "wikipedia":
                        interwiki = langPortion + ":";
                        break;
                    case "wiktionary":
                        interwiki = "wikt:" + langPortion + ":";
                        break;
                    case "wikibooks":
                        interwiki = "b:" + langPortion + ":";
                        break;
                    case "wikinews":
                        interwiki = "n:" + langPortion + ":";
                        break;
                    case "wikisource":
                        interwiki = "s:" + langPortion + ":";
                        break;
                    case "wikiquote":
                        interwiki = "q:" + langPortion + ":";
                        break;
                    case "wikiversity":
                        interwiki = "v:" + langPortion + ":";
                        break;
                    default:
                        throw new Exception((String)Program.msgs["20004"]);
                }
            }

            if (this.ContainsKey(projectName))
                throw new Exception(Program.GetFormatMessage(16400, projectName));

            logger.InfoFormat("Registering new project {0} with interwiki {1}", projectName, interwiki);
            Project prj = new Project();
            prj.projectName = projectName;
            prj.interwikiLink = interwiki;
            switch(projectName){
                case "mediawiki.wikipedia":
                    prj.langCode = "en";
                    prj.rooturl = "https://www.mediawiki.org/";
                    break;
                case "outreach.wikipedia":
                    prj.langCode = "en";
                    prj.rooturl = "https://outreach.wikimedia.org/";
                    break;
                case "testwikidata.wikipedia":
                    prj.langCode = "en";
                    prj.rooturl = "https://test.wikidata.org/";
                    break;
                case "wikidata.wikipedia":
                    prj.langCode = "en";
                    prj.rooturl = "https://www.wikidata.org/";
                    break;
                default:
                    prj.langCode = lang;
                    prj.rooturl = "https://" + projectName + ".org/";
                    break;
            }
            prj.RetrieveWikiDetails();
            this.Add(projectName, prj);
            // Join the new channel
            logger.InfoFormat("Joining RCReader channel: #{0}", projectName);
            Program.rcirc.rcirc.RfcJoin("#" + projectName);

            // Dump new settings
            DumpToFile();
        }

        /// <summary>
        /// Removes a project from the ProjectList
        /// </summary>
        /// <param name="projectName">Name of the project to remove</param>
        public void DeleteProject(string projectName)
        {
            if (!this.ContainsKey(projectName))
            {
                throw new Exception(Program.GetFormatMessage(16401, projectName));
            }

            logger.Info("Deleting existing project " + projectName);

            // Leave monitoring channel:
            logger.Info("Leaving #" + projectName);
            Program.rcirc.rcirc.RfcPart("#" + projectName, "No longer monitored");

            // Wait for existing RCEvents in separate thread to go through:
            Thread.Sleep(4000);

            // Finally, remove from list:
            this.Remove(projectName);

            // Dump new settings:
            DumpToFile();
        }

        public void ReloadAllWikis()
        {
            Thread.CurrentThread.Name = "ReloadAll";

            Program.SendMessageF(Meebey.SmartIrc4net.SendType.Message, currentBatchReloadChannel,
                                 "Request to reload all " + this.Count.ToString() + " wikis accepted.",
                                 Meebey.SmartIrc4net.Priority.High);

            foreach (DictionaryEntry dicent in this)
            {
                Project prj = (Project)dicent.Value;
                prj.RetrieveWikiDetails();
                Thread.Sleep(600);
            }

            // Dump new settings:
            DumpToFile();

            Program.SendMessageF(Meebey.SmartIrc4net.SendType.Message, currentBatchReloadChannel,
                                 "Reloaded all wikis. Phew, give the Wikimedia servers a break :(",
                                 Meebey.SmartIrc4net.Priority.High);
        }
    }
}

File: cvnbot/projectlist.py
import logging
import time
import xml.etree.ElementTree as ElementTree

from .ircclient import Priority, SendType
from .project import Project

logger = logging.getLogger("CVNBot.ProjectList")


class ProjectList:
    """The list of monitored wikis, stored as Projects.xml."""

    def __init__(self, program):
        self.bot = program
        self.projects = {}
        self.fn_projects_xml = ""

    # -- Map interface ----------------------------------------------------

    def __contains__(self, project_name):
        return project_name in self.projects

    def __getitem__(self, project_name):
        return self.projects[project_name]

    def __iter__(self):
        return iter(self.keys())

    def __len__(self):
        return len(self.projects)

    def keys(self):
        return sorted(self.projects)

    def values(self):
        return [self.projects[name] for name in self.keys()]

    def get(self, project_name, default=None):
        return self.projects.get(project_name, default)

    # -- Persistence ------------------------------------------------------

    def dump_to_file(self):
        """Save all project details to Projects.xml."""
        logger.info("Saving configuration to %s", self.fn_projects_xml)

        with open(self.fn_projects_xml, "w", encoding="utf-8") as handle:
            handle.write("<projects>\n")
            for project in self.values():
                # Get each Project's details and append it to the XML file
                handle.write(project.dump_project_details() + "\n")
            handle.write("</projects>\n")

    def load_from_file(self):
        try:
            root = ElementTree.parse(self.fn_projects_xml).getroot()
        except FileNotFoundError:
            logger.info("Creating new projects file at %s", self.fn_projects_xml)
            self.dump_to_file()
            return

        logger.info("Reading projects from %s", self.fn_projects_xml)
        for element in root:
            project = Project()
            project.read_project_details(element)
            self.projects[project.project_name] = project

    def add_new_project(self, project_name, interwiki):
        """
        Download new project details, start monitoring, and sync to Projects.xml.

        Args:
            string project_name: Name of the project e.g., "en.wikipedia"
            string interwiki: Interwiki link, e.g., "it:s:", can be empty string
                              in which case it is guessed from the project name,
                              e.g. "en.wikisource" -> "s:en:"
                              and "nl.wikipedia" -> "nl:"
        """
        if interwiki == "":
            # Try to guess interwiki
            if "." not in project_name:
                # Cannot guess; probably something like "mediawiki"
                raise Exception(self.bot.msgs["20004"])
            lang_portion, proj_portion = project_name.split(".", 1)
            # Interwiki prefixes per project family
            INTERWIKI_PREFIXES = {
                "wikipedia": "",
                "wiktionary": "wikt:",
                "wikibooks": "b:",
                "wikinews": "n:",
                "wikisource": "s:",
                "wikiquote": "q:",
                "wikiversity": "v:",
            }
            if proj_portion not in INTERWIKI_PREFIXES:
                raise Exception(self.bot.msgs["20004"])
            interwiki = INTERWIKI_PREFIXES[proj_portion] + lang_portion + ":"

        if project_name in self.projects:
            raise Exception(self.bot.msgs.format(16400, project_name))

        logger.info(
            "Registering new project %s with interwiki %s", project_name, interwiki
        )
        project = Project()
        project.project_name = project_name
        project.interwiki_link = interwiki

        # Wikis whose project name does not match their domain name
        SPECIAL_ROOT_URLS = {
            "mediawiki.wikipedia": "https://www.mediawiki.org/",
            "outreach.wikipedia": "https://outreach.wikimedia.org/",
            "testwikidata.wikipedia": "https://test.wikidata.org/",
            "wikidata.wikipedia": "https://www.wikidata.org/",
        }
        project.rooturl = SPECIAL_ROOT_URLS.get(project_name, "https://{0}.org/".format(project_name))
        project.retrieve_wiki_details()
        self.projects[project_name] = project

        # Join the new channel
        logger.info("Joining RCReader channel: #%s", project_name)
        self.bot.rcreader.rcirc.rfc_join("#" + project_name)

        self.dump_to_file()

    def delete_project(self, project_name):
        """Remove a project from the list, stop monitoring, and sync to Projects.xml."""
        if project_name not in self.projects:
            raise Exception(self.bot.msgs.format(16401, project_name))

        logger.info("Deleting existing project %s", project_name)

        # Leave monitoring channel
        logger.info("Leaving #%s", project_name)
        self.bot.rcreader.rcirc.rfc_part("#" + project_name, "No longer monitored")

        # Wait for existing RCEvents in separate thread to go through
        time.sleep(4)

        # Finally, remove from list
        del self.projects[project_name]

        self.dump_to_file()

    def reload_all_wikis(self, origin_channel):
        """
        Redownload the details of all projects and sync to Projects.xml.

        This will block for a long time and should run in its own thread.
        """
        self.bot.send_message(
            SendType.MESSAGE,
            origin_channel,
            "Request to reload all {0} wikis accepted.".format(len(self)),
            Priority.HIGH,
        )

        for project in self.values():
            project.retrieve_wiki_details()
            time.sleep(0.6)

        self.dump_to_file()

        self.bot.send_message(
            SendType.MESSAGE,
            origin_channel,
            "Reloaded all wikis. Phew, give the Wikimedia servers a break :(",
            Priority.HIGH,
        )

    # -- Namespaces -------------------------------------------------------

    def translate_namespace(self, project_name, original_title):
        """Translate a local title's namespace to canonical English."""
        project = self.projects.get(project_name)
        if project is None:
            return original_title
        return project.translate_namespace(original_title)
